app/models/office.rb

```diff
@@ -57,21 +65,41 @@ def full_street_address
     [address_line_one, address_line_two, address_postcode, 'United Kingdom'].reject(&:blank?).join(', ')
   end
 
-  private
+  def geocode
+    return false unless valid?
+    return true unless needs_to_be_geocoded?
 
-  def upcase_postcode
-    address_postcode.upcase! if address_postcode.present?
+    self.coordinates = ModelGeocoder.geocode(self)
+    add_geocoding_failed_error unless geocoded?
+
+    geocoded?
   end
 
-  def geocode_and_reindex_firm
-    return if destroyed?
-    if valid? and main_office?
-      firm.geocode_and_reindex # until we move the geocoding to offices, geocode the firm if this is the main office
-    end
+  def needs_to_be_geocoded?
+    !geocoded? || has_address_changes?
   end
 
-  def main_office?
-    firm.try(:main_office) == self
+  def has_address_changes?
+    ADDRESS_FIELDS.any? { |field| changed_attributes.include? field }
+  end
+
+  def save_with_geocoding
+    geocode && save
+  end
+
+  def update_with_geocoding(office_params)
+    self.attributes = office_params
+    save_with_geocoding
+  end
+
+  private
+
+  def add_geocoding_failed_error
+    errors.add(:address, 'could not be geocoded')
```

Comment:
Also note this hard coded error message will be replaced by a translation in an upcoming PR.